File: zors_pos_backend/models/Product.ts
import mongoose, { Document, Schema } from "mongoose";

export interface IProduct extends Document {
  shortId: string;
  name: string;
  description?: string;
  category: string;
  costPrice: number;
  sellingPrice: number;
  stock: number;
  minStock: number;
  barcode?: string;
  image?: string;
  supplier?: string;
  discount?: number;
  size?: string;
  dryfood?: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const ProductSchema = new Schema<IProduct>({
  shortId: { type: String },
  name: { type: String, required: true },
  description: { type: String },
  category: { type: String, required: true },
  costPrice: { type: Number, required: true },
  sellingPrice: { type: Number, required: true },
  stock: { type: Number, required: true, default: 0 },
  minStock: { type: Number, required: true, default: 5 },
  barcode: { type: String, unique: true, sparse: true },
  image: { type: String },
  supplier: { type: String },
  discount: { type: Number, default: 0 },
  size: { type: String },
  dryfood: { type: Boolean, default: false },
}, {
  timestamps: true,
});

// Generate unique 6-digit ID
ProductSchema.pre('save', async function() {
  if (!this.shortId) {
    let uniqueId: string;
    let isUnique = false;
    
    while (!isUnique) {
      // Generate random 6-digit number
      uniqueId = Math.floor(100000 + Math.random() * 900000).toString();
      
      // Check if shortId already exists (not _id)
      const existing = await mongoose.models.Product.findOne({ shortId: uniqueId });
      if (!existing) {
        isUnique = true;
        this.shortId = uniqueId;
      }
    }
  }
});

export default mongoose.models.Product || mongoose.model<IProduct>('Product', ProductSchema);

File: zors_pos_mobile/lib/models/product.dart
class Product {
  final String id;
  final String name;
  final String? description;
  final String category;
  final double costPrice;
  final double sellingPrice;
  final int stock;
  final int minStock;
  final String? barcode;
  final String? image;
  final String? supplier;
  final double? discount;
  final String? size;
  final bool? dryfood;
  final DateTime createdAt;
  final DateTime updatedAt;

  Product({
    required this.id,
    required this.name,
    this.description,
    required this.category,
    required this.costPrice,
    required this.sellingPrice,
    required this.stock,
    required this.minStock,
    this.barcode,
    this.image,
    this.supplier,
    this.discount,
    this.size,
    this.dryfood,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['_id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'],
      category: json['category'] ?? '',
      costPrice: (json['costPrice'] ?? 0).toDouble(),
      sellingPrice: (json['sellingPrice'] ?? 0).toDouble(),
      stock: json['stock'] ?? 0,
      minStock: json['minStock'] ?? 5,
      barcode: json['barcode'],
      image: json['image'],
      supplier: json['supplier'],
      discount: (json['discount'] ?? 0).toDouble(),
      size: json['size'],
      dryfood: json['dryfood'] ?? false,
      createdAt: DateTime.parse(
        json['createdAt'] ?? DateTime.now().toIso8601String(),
      ),
      updatedAt: DateTime.parse(
        json['updatedAt'] ?? DateTime.now().toIso8601String(),
      ),
    );
  }

  Map<String, dynamic> toJson() {
    final map = <String, dynamic>{
      'name': name,
      'description': description,
      'category': category,
      'costPrice': costPrice,
      'sellingPrice': sellingPrice,
      'stock': stock,
      'minStock': minStock,
      'barcode': barcode,
      'image': image,
      'supplier': supplier,
      'discount': discount,
      'size': size,
      'dryfood': dryfood,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
    };
    // Only include _id when present; let server generate otherwise
    if (id.isNotEmpty) {
      map['_id'] = id;
    }
    return map;
  }
}
